File: sort/mergeSort/merge_sort.rb
def merge(arr1, arr2)
    result = []
    i = 0
    j = 0

    while (i < arr1.size and j < arr2.size)
        if arr1[i] < arr2[j]
            result.push(arr1[i])
            i+=1
        else
            result.push(arr2[j])
            j+=1
        end
    end

    while i < arr1.size
        result.push(arr1[i])
        i+=1
    end

    while j < arr2.size
        result.push(arr1[j])
        j+=1
    end
    return result
end

def merge_sort(arr)
    len = arr.size
    return arr if len <= 1

    middle = len / 2
    left = merge_sort(arr.slice(0, middle))
    right = merge_sort(arr.slice(middle, len))

    return merge(left, right)
end

print(merge_sort([12,6,4,1]))

File: challenges/mergeSort.js
function merge(arr1, arr2, callback) {
  if (typeof callback !== 'function') callback = (a, b) => a - b;

  let sortedAry = [];
  let i = 0;
  let j = 0;

  while (i < arr1.length && j < arr2.length) {
    let greaterThan = callback(arr1[i], arr2[j]);
    if (greaterThan >= 0) {
      sortedAry.push(arr2[j]);
      j++;
    } else {
      sortedAry.push(arr1[i]);
      i++;
    }
  }

  while (i < arr1.length) {
    sortedAry.push(arr1[i]);
    i++;
  }

  while (j < arr2.length) {
    sortedAry.push(arr2[j]);
    j++;
  }

  return sortedAry;
}

function mergeSort(arr, callback) {
  let middle = Math.floor(arr.length / 2);
  if (arr.length <= 1) return arr;

  let left = mergeSort(arr.slice(0, middle), callback);
  let right = mergeSort(arr.slice(middle), callback);

  return merge(left, right, callback);
}
